src/services/codeFixProvider.ts

```diff
@@ -3,6 +3,14 @@ namespace ts {
     export interface CodeFix {
         errorCodes: number[];
         getCodeActions(context: CodeFixContext): CodeAction[] | undefined;
+        createCodeFixDiagnosticIfApplicable?(node: Node, context: CodeFixDiagnoseContext): Diagnostic | undefined;

```

Comment:
We also want to use the same mechanism for refactorings. a refactoring should have isApplicable, and getCodeActions. I would recommend switching this into `isApplicable`, and an optional diagnostic message.